fms/modules/embedding.py

```diff
@@ -1,21 +1,123 @@
-    apply_rowwise_tp,
     copy_to_tensor_model_parallel_region,
-    reduce_from_tensor_model_parallel_region,
 )
 
 
+class AbsolutePositionEmbedding(nn.Module):
+    """Special form of embedding that includes a position encoding"""
+
+    def __init__(
+        self,
+        vocab_size: int,
+        emb_dim: int,
+        padding_idx: Optional[int] = None,
+        max_pos: int = 512,
+    ):
+        """
+        Initialize an AbsolutePositionEmbedding
+
+        Parameters
+        ----------
+        vocab_size: int
+            the length of the vocabulary
+        emb_dim: int
+            dimensionality of latent space
+        padding_idx: int, optional
+            Padding token index in the vocabulary. Typically, the token for
+            which positions will be zeroed out. If negative or None, no positions
+            will be zeroed. (default is None)
+        max_pos: int
+            the maximum possible sequence length (default is 512)
+        """
+        super().__init__()
+        self.max_pos = max_pos
+        self.emb_dim = emb_dim
+        self.padding_idx = padding_idx
+
+        if self.padding_idx is None:
+            self.emb = nn.Embedding(vocab_size, self.emb_dim)
+        else:
+            self.emb = nn.Embedding(
+                vocab_size, self.emb_dim, padding_idx=self.padding_idx
+            )
+
+        self.pos_emb = nn.Embedding(max_pos, self.emb_dim)
+        self.register_buffer("pos_id", torch.arange(max_pos).unsqueeze(0))
+
+    def reset_params(self):
+        # Defaults to norm-preserving in reverse op, unit vector in forward op
+        layers = ["emb"]
+        layers.append("pos_emb")
+        for layer in layers:
+            nn.init.trunc_normal_(
+                getattr(self, layer).weight, mean=0.0, std=self.emb_dim**-0.5
+            )
+        # Preserve pad index dummy-hood
+        if self.padding_idx is not None:
+            self.emb.weight.data[self.padding_idx].zero_()
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        position_ids: Optional[torch.LongTensor] = None,
+        correct_pads: bool = False,
+    ):
+        """
+        perform a forward pass of absolute positional embedding
+
+        Parameters
+        ----------
+        x: torch.Tensor
+            the input tensor
+        position_ids: torch.LongTensor, optional
+            a tensor which signifies the positions in the position embedding. If
+            None, will simply be a range from beginning to end of sequence for
+            each sequence in the batch (default is None)
+        correct_pads: bool
+            if True, will assume position_ids has not been corrected for padding
+            and will perform the necessary shifting, otherwise assume
+            position_ids has already been corrected (default is False)
+
+        Returns
+        -------
+        torch.Tensor
+            the output of absolute positional embeddings
+        """
+        x_emb = self.emb(x)
```

Comment:
is this still doing both word and position embeddings?

Reply:
> is this still doing both word and position embeddings?

Ah yes, it includes both the embedding and the positional embedding. Do you suggest I completely separate them, and assume the x that comes in has already been run through the word embedding?

Reply:
I thought that was the goal of this PR -- to break out position embedding from word embeddings? if the new class does both then... I'm not clear what the difference is.

Reply:
The old class had the option to do word embeddings, optionally do positional embeddings, optionally do lm head. I was trying to make this clearer. This removes the need to choose abs_pos=True/False or reversible=True/False. 

Having said that, we could take this one step further and just completely separate the word embedding from the absolute positional embedding (The way I saw it when I implemented this was, absolute positional embedding will always require the initial word embedding, so why not just have it do both in this module - basically AbsolutePositionEmbedding => embedding + position encoding)

Reply:
I would completely separate both in different classes, and just have the two calls. Most current models don't use this kind of embeddings anymore tbh, so for BigCode and Roberta we can have an extra line, while keeping the rest of the code for the llama family and T5 cleaner (and probably any future models we develop in-house).

Reply:
It is not really an extra line, since keeping them together allows for us to take in the original input and do all of the logic in the AbsolutePositionEmbedding for the position_ids. In this case, I will need to add all of that logic to the model, which will probably be about 20 or so lines. If we don't see this ever being re-used, I dont mind putting it in, but it will at least be re-used here and with Roberta.

Reply:
> which will probably be about 20 or so lines

I wouldn't just copy/paste back into the model. Part of the complexity comes from this being a separate class.

E.g. you had to pass `correct_pads=position_ids is None`. If you were still in the model code, you would have already corrected pads in the `is None` path while generating the position_ids. If that code was something like `(x != pad_idx).cumsum(1)` you also wouldn't need to then `clamp(min=0)` afterwards.